Kernel/TTY/TTY.cpp

```diff
@@ -395,7 +399,26 @@ void TTY::hang_up()
 
 ProcessGroupID TTY::pgid() const
 {
-    return m_process ? m_process->pgid() : 0;
+    if (m_process)
+        return m_process->pgid();
+
+    // Try to find a new process in the same pgid.
+    if (m_last_assigned_pgid != 0) {
+        InterruptDisabler disabler;
+        Process::for_each_in_pgrp(m_last_assigned_pgid, [&](auto& process) -> IterationDecision {
+            m_process = process.make_weak_ptr();
+            return IterationDecision::Break;
+        });
+    }
```

Comment:
What do other systems do?

Reply:
Seems to me that at least linux doesn't care which process owns the tty, it just checks for the pgrp.
and `!pgrp` just makes it allow reads/writes:
```c
        pgrp = task_pgrp(current);

	tty_pgrp = tty->pgrp;
	
	if (tty_pgrp && pgrp != tty_pgrp) {
		if (is_ignored(sig)) {
			if (sig == SIGTTIN)
				ret = -EIO;
		} else if (is_current_pgrp_orphaned())
			ret = -EIO;
		else {
			kill_pgrp(pgrp, sig, 1);
			set_thread_flag(TIF_SIGPENDING);
			ret = -ERESTARTSYS;
		}
	}
```